src/privatesend/privatesend-server.cpp

```diff
@@ -628,48 +538,128 @@ bool CPrivateSendServer::IsInputScriptSigValid(const CTxIn& txin)
+        }
+        // IsPayToPublicKeyHash() above already checks for scriptPubKey size,
+        // no need to double check, hence no usage of ERR_NON_STANDARD_PUBKEY
+        return true;
+    };
+
+    for (const auto& txout : entry.vecTxOut) {
+        if (!checkTxOut(txout)) {
+            return false;
+        }
+    }
+
+    for (const auto& txin : entry.vecTxDSIn) {
+        LogPrint(BCLog::PRIVATESEND, "CPrivateSendServer::%s -- txin=%s\n", __func__, txin.ToString());
+
+        if (txin.prevout.IsNull()) {
+            LogPrint(BCLog::PRIVATESEND, "CPrivateSendServer::%s -- ERROR: invalid input!\n", __func__);
+            nMessageIDRet = ERR_INVALID_INPUT;
+            ConsumeCollateral(connman, entry.txCollateral);
+            return false;
+        }
+
         for (const auto& entry : vecEntries) {
             for (const auto& txdsin : entry.vecTxDSIn) {
                 if (txdsin.prevout == txin.prevout) {
-                    LogPrint(BCLog::PRIVATESEND, "CPrivateSendServer::AddEntry -- found in txin\n");
+                    LogPrint(BCLog::PRIVATESEND, "CPrivateSendServer::%s -- ERROR: already have this txin in entries\n", __func__);
                     nMessageIDRet = ERR_ALREADY_HAVE;
+                    // Two peers sent the same input? Can't really say who is the malicious one here,
+                    // could be that someone is picking someone else's inputs randomly trying to force
+                    // collateral consumption. Do not punish.
                     return false;
                 }
             }
         }
+
+        Coin coin;
+        CScript scriptPubKeyIn;
+        auto mempoolTx = mempool.get(txin.prevout.hash);
+        // We or the other peer could be simply out of sync in all three cases below, do not punish.
+        if (mempoolTx != nullptr) {
+            if (mempool.isSpent(txin.prevout) || !llmq::quorumInstantSendManager->IsLocked(txin.prevout.hash)) {
+                LogPrint(BCLog::PRIVATESEND, "CPrivateSendServer::%s -- ERROR: spent or non-locked mempool input! txin=%s\n", __func__, txin.ToString());
+                nMessageIDRet = ERR_MISSING_TX;
+                return false;
+            }
+            if (!checkTxOut(mempoolTx->vout[txin.prevout.n])) {
+                return false;
+            }
+            scriptPubKeyIn = mempoolTx->vout[txin.prevout.n].scriptPubKey;
+        } else if (GetUTXOCoin(txin.prevout, coin)) {
+            if (!checkTxOut(coin.out)) {
+                return false;
+            }
+            scriptPubKeyIn = coin.out.scriptPubKey;
+        } else {
+            LogPrint(BCLog::PRIVATESEND, "CPrivateSendServer::%s -- ERROR: missing input! txin=%s\n", __func__, txin.ToString());
+            nMessageIDRet = ERR_MISSING_TX;
+            return false;
+        }
+
+        // The same size and denom for inputs and outputs ensures their total value is also the same,
+        // no need to double check, hence no usage of ERR_FEES
```

Comment:
Although, yes this is the case... I would like an explicit check such that if we make changes (like allowing multiple types of denoms in one mix) in the future it would be harder to open up an avenue for attack.

Reply:
Actually, I'm trying to move in the exactly opposite direction and remove all possibilities for multiple denoms in one mixing tx - we already tried that before and that's where a lot of unnecessary complexity in PS comes from. We already know multiple denoms in one tx do not really work and you MUST use the same amounts. Next steps in PS refactoring is to have `GetDenominations`-> `GetDenomination` etc. which will further ensure that session denoms can only be single bits and not some combination of them.

Reply:
I completely agree that that is the correct direction to go... But I don't think it'd be bad to have an explicit check that total value is the same

Reply:
ok, I agree